mozjs/js/rust/src/jsslotconversions.rs

```diff
@@ -472,3 +474,44 @@ impl<T> ToFromJsSlots for Vec<T> where T: fmt::Debug {
     fn trace(_trc: *mut JSTracer, _obj: *mut JSObject, _offset: u32) {
     }
 }
+
+impl ToFromJsSlots for *mut JSString {
+    type Target = Self;
+    const NEEDS_FINALIZE: bool = false;
+    const NEEDS_TRACE: bool = false;
+
+    unsafe fn from_slots(object: *mut JSObject, cx: *mut JSContext, offset: u32)
+                         -> Self::Target {
+        rooted!(in(cx) let val0 = JS_GetReservedSlot(object, offset));
+        if val0.is_undefined() {
+            return 0 as *mut _;
+        }
+        let conversion0 =
+            FromJSValConvertible::from_jsval(cx, val0.handle(), ())
+            .expect("Should never put anything into a JS slot that we can't \
+                     convert back out again");
+        rooted!(in(cx) let undef_val = jsval::UndefinedValue());
+        JS_SetReservedSlot(object, offset, &*undef_val);
+        let ptr_val: JS::Value = match conversion0 {
+            ConversionResult::Success(v) => v,
+            ConversionResult::Failure(why) => {
+                panic!("Should never put anything into a JS slot that we \
+                        can't convert back out again: {}",
+                       why);
+            }
+        };
+        let ptr = ptr_val.to_string();
+        ptr
```

Comment:
Since `*mut JSString` is `Copy`, we shouldn't need to do the take/replace dance here. Instead, something like:

```rust
unsafe fn from_slots(object: *mut JSObject, cx: *mut JSContext, offset: u32) -> Self::Target {
    rooted!(in(cx) let val = JS_GetReservedSlot(object, offset);
    if val.is_string() {
        return val.to_string();
    } else if val.is_null() {
        return ptr::null_mut();
    } else {
        panic!("<*mut JSString as ToFromJsSlots>::from_slots called on non-string/null slot");
    }
}
```